src/test/bin/winkernel/control.cpp

```diff
@@ -65,6 +65,113 @@ WDFDEVICE QuicTestCtlDevice = nullptr;
 
+typedef struct QuicTestNmrClient {
+    NPI_CLIENT_CHARACTERISTICS NpiClientCharacteristics;
+    HANDLE NmrClientHandle;
+    NPI_MODULEID ModuleId;
+    CXPLAT_EVENT RegistrationCompleteEvent;
+    MSQUIC_NMR_DISPATCH* ProviderDispatch;
+    BOOLEAN Deleting;
+} QuicTestNmrClient;
+
+static QuicTestNmrClient NmrClient;
+
+static
+NTSTATUS
+QuicTestClientAttachProvider(
+    _In_ HANDLE NmrBindingHandle,
+    _In_ void *ClientContext,
+    _In_ const NPI_REGISTRATION_INSTANCE *ProviderRegistrationInstance
+    )
+{
+    UNREFERENCED_PARAMETER(ProviderRegistrationInstance);
+
+    NTSTATUS Status;
+    QuicTestNmrClient* Client = (QuicTestNmrClient*)ClientContext;
+    void* ProviderContext;
+
+    #pragma warning(suppress:6387) // _Param_(2) could be '0' - by design.
+    Status =
+        NmrClientAttachProvider(
+            NmrBindingHandle,
+            Client,
+            NULL,
+            &ProviderContext,
+            (const void**)&Client->ProviderDispatch);
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "NmrClientAttachProvider failed");
+    }
+    return Status;
+}
+
+static
+NTSTATUS
+QuicTestClientDetachProvider(
+    _In_ void *ClientBindingContext
+    )
+{
+    QuicTestNmrClient* Client = (QuicTestNmrClient*)ClientBindingContext;
+    if (InterlockedFetchAndSetBoolean(&Client->Deleting)) {
+        return STATUS_SUCCESS;
+    } else {
+        return STATUS_PENDING;
+    }
+}
+
+NTSTATUS
+QuicTestRegisterNmrClient(
+    void
+    )
+{
+    NPI_REGISTRATION_INSTANCE *ClientRegistrationInstance;
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    CxPlatEventInitialize(&NmrClient.RegistrationCompleteEvent, FALSE, FALSE);
+    NmrClient.ModuleId.Length = sizeof(NmrClient.ModuleId);
+    NmrClient.ModuleId.Type = MIT_GUID;
+    NmrClient.ModuleId.Guid = MSQUIC_MODULE_ID;
+
+    NmrClient.NpiClientCharacteristics.Length = sizeof(NmrClient.NpiClientCharacteristics);
+    NmrClient.NpiClientCharacteristics.ClientAttachProvider = QuicTestClientAttachProvider;
+    NmrClient.NpiClientCharacteristics.ClientDetachProvider = QuicTestClientDetachProvider;
+
+    ClientRegistrationInstance = &NmrClient.NpiClientCharacteristics.ClientRegistrationInstance;
+    ClientRegistrationInstance->Size = sizeof(*ClientRegistrationInstance);
+    ClientRegistrationInstance->Version = 0;
+    ClientRegistrationInstance->NpiId = &MSQUIC_NPI_ID;
+    ClientRegistrationInstance->ModuleId = &NmrClient.ModuleId;
+
+    Status =
+        NmrRegisterClient(
+            &NmrClient.NpiClientCharacteristics, &NmrClient, &NmrClient.NmrClientHandle);
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "NmrRegisterClient failed");
+        goto Exit;
+    }
+
+    if (!CxPlatEventWaitWithTimeout(NmrClient.RegistrationCompleteEvent, 1000)) {
```

Comment:
It's unclear to me where/who sets this.

Reply:
It's supposed to be set in at the end of clientAttachProvider callback. I am surprised the code still works.